onnxruntime/python/onnxruntime_pybind_mlvalue.cc

```diff
@@ -118,110 +118,180 @@ const DataTypeImpl* NumpyToOnnxRuntimeTensorType(int numpy_type) {
   }
 }
 
+TensorShape GetArrayShape(PyArrayObject* pyObject) {
+  int ndim = PyArray_NDIM(pyObject);
+  const npy_intp* npy_dims = PyArray_DIMS(pyObject);
+  std::vector<int64_t> dims(ndim);
+  for (int i = 0; i < ndim; ++i) {
+    dims[i] = npy_dims[i];
+  }
+  TensorShape shape(std::move(dims));
+  return shape;
+}
+
+// This is a one time use, ad-hoc allocator that allows Tensors to take ownership of
+// python array objects and use the underlying memory directly and
+// properly deallocated them when they are done.
+//
+// This addresses the case when our interfaces receive python lists on the input.
+// We have to convert them into new Numpy arrays which 1) needs to be properly deallocated
+// because they are not owned by the calling python code (we create it inside pybind code)
+// 2) we still want to avoid yet another data copy and use it directly if possible
+// 3) string data types still need to be copied
+//
+// This is a stateful allocator. It will always return the same pre-allocated
+// buffer pointer and will own references to underlying objects.
+class OrtPybindAllocator : public IAllocator {
+ public:
+   // This constructor is used when we create numpy array from python list
+  OrtPybindAllocator(PyArrayObject* pyObject, const std::string& value_name, const OrtMemoryInfo& mem_info)
+      : pyObject_(pyObject, DecRefFn<PyArrayObject>()),
+        pyObjectContinuous_(PyArray_GETCONTIGUOUS(pyObject), DecRefFn<PyArrayObject>()),
+        mem_info_(mem_info) {
+    ORT_ENFORCE(pyObjectContinuous_ != nullptr, "The object must be a contiguous array for input :", value_name);
+  }
+
+  // Constructor to use when a continuous array had to be copied. Instead of creating yet another copy
+  // we are still able to use it directly for primitive types
+  OrtPybindAllocator(UniqueDecRefPtr<PyArrayObject>&& pyContinuous, const std::string& value_name, const OrtMemoryInfo& mem_info) 
+    : pyObject_(nullptr, DecRefFn<PyArrayObject>()),
+      pyObjectContinuous_(std::move(pyContinuous)),
+      mem_info_(mem_info){
+    ORT_ENFORCE(pyObjectContinuous_ != nullptr, "Expecting a valid contiguous array:", value_name);
+  }
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(OrtPybindAllocator);
+
```

Comment:
Does this need 'move' semantics to be supported?

Reply:
I think the opposite since the Allocator semantics is always a shared_ptr<> we move that.

---
In reply to: [434975882](https://github.com/microsoft/onnxruntime/pull/4124#discussion_r434975882) [](ancestors = 434975882)